src/lib.rs

```diff
@@ -428,6 +453,20 @@ pub trait Perform {
     /// The `ignore` flag indicates that more than two intermediates arrived and
     /// subsequent characters were ignored.
     fn esc_dispatch(&mut self, _intermediates: &[u8], _ignore: bool, _byte: u8) {}
+
+    /// The start of an opaque sequence (SOS, PM or APC) has been detected.
+    ///
+    /// The `kind` parameter indicates the type of sequence that was started.
+    ///
+    /// Until the opaque sequence ends (at which point `opaque_end` will be called), invalid
+    /// characters will be ignored while valid characters will be passed on to `opaque_put`.
+    fn opaque_start(&mut self, _kind: OpaqueSequenceKind) {}
+
+    /// We've reached the end of the ongoing opaque sequence (SOS, PM or APC).
+    fn opaque_end(&mut self) {}
+
+    /// A byte has been received as part of an ongoing opaque sequence.
+    fn opaque_put(&mut self, _byte: u8) {}
```

Comment:
Is there a good reason why you decided to make this one set of functions with a `_kind` parameter, instead of one for each kind?

Reply:
There are several reasons for this:

1. I wanted to minimize the amount of packed actions that I would need to add, since the current parser only supports 15 packed actions. If I separate these functions, we will need to track states for APC, SOS and PM sequences separately and then `Action::OpaquePut` would need to be packed, so this would consume two extra packed action slots. I wanted to minimize future performance impact for parsers that don't need to deal with these sequences, so I opted to unify all opaque sequence handling.
(`opaque_start` and `opaque_end` won't consume a slot, but I thought it's better to keep the names consistent)

2. SOS and PM sequences are even more rare. I've only found mention of SOS used as a C1 control code for MARC-21 Unicode variant. It is not even used in the context of a terminal or a terminal multiplexer.
PM is used by [screen](https://www.gnu.org/software/screen/manual/html_node/Privacy-Message.html) for displaying a notification and by [jexer](https://gitlab.com/AutumnMeowMeow/jexer/-/wikis/jexer-sequences#nested-text-mouse-pointer) for handling nested mouse pointers. I wasn't sure if it's a good idea to dedicate separate functions for these rare use cases.

3. APC, SOS and PM all work the same way[^1]. They all take opaque strings that are terminated with ST and they all don't define any semantics for how these strings should work. Since there is no well-defined structure, the implementation needs to create its own parser and track its own state for these strings anyway. In that case, separating APC, SOS and PM to different sets of functions doesn't help the application very much.

---

That's more or less my rationale, but I'm open to hear what you think about it.

[^1]: I'm stretching the truth a bit: According to the ECMA-48 spec, SOS allows any character to be embedded (except for escape codes or C1 sequences representing SOS or ST), while PM and APC only allow the ranges 0x08-0x0d, 0x20-0x7e. In practice, since nobody (as far as I know) is using SOS sequences as far as we know, it's quite hard for me to know how to implement it properly.

Reply:
If you pass them as an enum argument to this function, then this means you need to identify them anyway. It just seems like a way to avoid doing it properly and if we're not doing it properly what's the point? A type of escape sequence should either be **properly** supported, or not at all.

Reply:
>  If you pass them as an enum argument to this function, then this means you need to identify them anyway.

This was my third point. You need to identify the actual (private) sequence type that APC/PM/SOS is used for anyway, even if I separate the function.

But I understand that you see separating them out as more proper.

What would you expect then? Three sets of functions and adding the necessary extra packed states?

Reply:
I would expect three sets of functions. How that is done is an implementation detail irrelevant to the consumer (as long as it's fast).